website/mailing_list/utils.py

```diff
@@ -0,0 +1,426 @@
+    update_multiple_users_in_list(*args, **kwargs)
+
+@queued_task
+@app.task(bind=True, max_retries=3, default_retry_delay=3 * 60)  # Retry after 3 minutes
+def celery_full_update(*args, **kwargs):
+    full_update(*args, **kwargs)
+
+
+###############################################################################
+# Signalled Functions
+###############################################################################
+
+@contributor_added.connect
+def subscribe_contributor_to_mailing_list(node, contributor, auth=None):
+    if node.mailing_enabled and contributor.is_active:
+        subscription = node.get_or_create_mailing_list_subscription()
+        subscription.add_user_to_subscription(contributor, 'email_transactional', save=True)
+        celery_update_single_user_in_list(node._id, contributor._id)
+        node.mailing_updated = True
+        node.save()
+
+@contributor_removed.connect
+def unsubscribe_contributor_from_mailing_list(node, user, auth=None):
+    if node.mailing_enabled:
+        subscription = node.get_or_create_mailing_list_subscription()
+    else:
+        from website.models import NotificationSubscription  # avoid circular import
+        subscription = NotificationSubscription.load(to_subscription_key(node._id, 'mailing_list_events'))
+    if subscription:
+        subscription.remove_user_from_subscription(user)
+        node.mailing_updated = True
+        node.save()
+    celery_remove_user_from_list(node._id, user._id)
+
+
+@user_confirmed.connect
+def resubscribe_on_confirm(user):
+    for node in user.contributed:
+        subscribe_contributor_to_mailing_list(node, user)
+
+@node_deleted.connect
+def remove_list_for_deleted_node(node):
+    node.mailing_updated = True
+    node.save()
+    celery_delete_list(node._id)
+
+
+###############################################################################
+# Mailing List Helper Functions
+###############################################################################
+
+def jsonify_users_list(users, unsubs=(), initial=False):
+    """ Serializes and paginates list of users for Mailgun
+
+    :param list users: users to serialize
+    :param list unsubs: unsubscribed users
+    :param bool initial: whether or not to insert mailing_list_robot
+    """
+    members_list = []
+    if initial:
+        members_list.append({'address': address('mailing_list_robot'), 'subscribed': True})  # Routing robot
+
+    for member in users:
+        for email in member.emails:
+            members_list.append({
+                'address': email,
+                'subscribed': member not in unsubs and email == member.username,
+                'vars': {'_id': member._id, 'primary': email == member.username}
+            })
+
+    # Mailgun allows 1000 user upserts per request
+    return [members_list[i:i + 999] for i in range(0, len(members_list), 999)]
+
+def address(node_id):
+    return '{}@{}'.format(node_id, furl(settings.DOMAIN).host)
```

Comment:
Per discussion: All mailgun configuration should go in the local settings file, so that changes made to the code do not override required mailgun configuration.